.gitlab-ci.yml

```diff
@@ -58,14 +59,34 @@ workflow:
 
 ### stage:                         test
 
-build-test:
+test:
   stage:                           test
   <<:                              *docker-env
+  script:
+    - echo cargo test --all
+
+build-linux:
+  stage:                           build
+  <<:                              *docker-env
+  <<:                              *collect-artifacts
+  rules:
+    - if: $CI_PIPELINE_SOURCE == "web"
+    - if: $CI_PIPELINE_SOURCE == "schedule"
+    - if: $CI_COMMIT_REF_NAME == "tags"
+    - if: $CI_COMMIT_REF_NAME =~ /^v[0-9]+\.[0-9]+.*$/        # i.e. v1.0, v2.1rc1
+    - if: $CI_COMMIT_REF_NAME == "master"
+      when: manual
+      allow_failure: true
+    - if: $CI_COMMIT_REF_NAME =~ /^[0-9]+$/                   # PRs
+      when: manual
+      allow_failure: true
   script:
     - cargo build --release
-    - cargo test --release --all
```

Comment:
what was wrong with keeping a test here? You don't even add additional flags, so it's perfect to build and test in the same job while reusing the cache.
And the mac build you can place after this job.
And publish job would only go after mac job and publish everything if the mac job is a success.